domain/src/main/java/org/oppia/android/domain/question/QuestionAssessmentCalculation.kt

```diff
@@ -0,0 +1,105 @@
+package org.oppia.android.domain.question
+
+import org.oppia.android.app.model.FractionGrade
+import org.oppia.android.app.model.UserAssessmentPerformance
+import javax.inject.Inject
+import javax.inject.Singleton
+
+/**
+ * Private mutable class that stores a grade as a fraction.
+ */
+private class MutableFractionGrade(var numerator: Int, var denominator: Int)
+
+/**
+ * Private class that computes the state of the user's performance at the end of a practice session.
+ * This class can exist across multiple training session instances, but calling code is responsible
+ * for ensuring it is properly reset.
+ */
+@Singleton
+internal
+class QuestionAssessmentCalculation @Inject constructor(
+  @ViewHintScorePenalty private val viewHintPenalty: Int,
+  @WrongAnswerScorePenalty private val wrongAnswerPenalty: Int,
+  @MaxScorePerQuestion private val maxScorePerQuestion: Int
+) {
+  private lateinit var skillIdList: List<String>
+  private lateinit var questionSessionMetrics: List<QuestionSessionMetrics>
+  private lateinit var totalScore: MutableFractionGrade
+  private lateinit var scorePerSkillMapping: MutableMap<String, MutableFractionGrade>
```

Comment:
I'm wondering how much of this class's state management could be simplified if it were a recreated object rather than being part of the Dagger graph. Consider instead creating a Factory class that's injected and that creates new instances of the calculation class. Then, you can pass in these values as part of the class's constructor and make them vals to have better immutability (which is a general design plus in the Java/Kotlin world). For a reference, take a look at UrlImageParser and PersistentCacheStore--both of those utilize injectable factories to create the underlying class.

I believe this also lets you get rid of the ``@Singleton`` qualifier, too, since no state needs to be stored in the Dagger graph then.

Reply:
Fixed! Thanks for the suggestion, I think this is definitely better